models/issue_dependency.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// IssueDependency represents an issue dependency
+type IssueDependency struct {
+	ID           int64          `xorm:"pk autoincr"`
+	UserID       int64          `xorm:"NOT NULL"`
+	IssueID      int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	DependencyID int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	CreatedUnix  util.TimeStamp `xorm:"INDEX created"`
+	UpdatedUnix  util.TimeStamp `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it aleready exists
+	err = issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return
+	}
+
+	// No error means the dependency doesn't exist nor is circular, so we can proceed to create a new one.
+	newIssueDependency := &IssueDependency{
+		UserID:       user.ID,
+		IssueID:      issue.ID,
+		DependencyID: dep.ID,
+	}
+
+	if _, err := sess.Insert(newIssueDependency); err != nil {
+		return err
+	}
+
+	// Add comment referencing the new dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+		return
+	}
+
+	// Create a new comment for the dependent issue
+	if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+		return
+	}
+
+	return sess.Commit()
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it exists
+	var exists bool
+	switch depType {
+	case DependencyTypeBlockedBy:
+		err = issueDepExists(sess, issue.ID, dep.ID)
+	case DependencyTypeBlocking:
+		err = issueDepExists(sess, dep.ID, issue.ID)
+	default:
+		return
+	}
+	if err != nil {
+		if IsErrDependencyExists(err) {
+			exists = true
+		} else {
+			return err
+		}
+	}
+
+	// If it exists, remove it
+	if exists {
```

Comment:
```
if !exists {
  return sess.Commit()
}
```

Reply:
Could you explain that? I don't understand why this would be necessary, because I still need to do all of that stuff when the dependency exists. Sure, I could add the piece of code you described, but I still need `sess.Commit()` after all the stuff which happens inside of that if (starting in line 97), which would mean to add another `sess.Commit()` after that (L 97).

IMHO the way I did it is ok, please tell me if I misunderstood you.

Reply:
I think it's about unnecessary nested code.

Reply:
But how is it unnecessary nested? :thinking:

Reply:
(I'll try to explain with "my" English) You should follow the "break early" rule and also keep standard/expected flow without indentation (not nested). Function name is `RemoveIssueDependency` so the expecting behaviour is to remove dependency. Missing issue is exception, not common flow/scenario.

If you know, you will not do anything else in some condition you should immidiately end processing. What if there will be more exceptions in future? Then you will have to create another nested block inside (and still not talking aboud returning different exceptions)... 

```
func shouldYouDoIt() {
    if !ex1 {
        if !ex2 {
           if !ex3 {
               doIt()
           }
        }
    }
}

# vs

func shouldYouDoIt() {
    if ex1 {
        return
    }

    if ex2 {
        return
    }

    if ex3 {
        return
    }

    doIt()
}
```

Another thing is, who will know later why that block is inside "if" and if he/she should add another code after it (between end of block and commit) or inside that block?

Reply:
I think it is better to have maybe more verbose but clear code. Also question is, if you rly need to commit something instead of returning error.

Reply:
Ok, I think I finally understood how it works. Thanks for taking the time to explain :)

I've fixed it. I also saw that we don't need to return `sess.Commit()` in any case, fixed that as well.